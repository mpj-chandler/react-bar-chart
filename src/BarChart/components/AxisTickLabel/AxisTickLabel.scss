
  .AxisTickLabel {
    font-size: calc(8px + 0.5vh);
    text-align: center;
  }