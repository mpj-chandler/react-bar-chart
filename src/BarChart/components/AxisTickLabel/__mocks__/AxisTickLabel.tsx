import React from 'react';

const AxisTickLabel: React.FC<any> = (props) => <div {...props}>MockAxisTickLabel {props.children}</div>;

export default AxisTickLabel;