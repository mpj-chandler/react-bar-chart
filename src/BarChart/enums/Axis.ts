enum Axis {
    XAxis,
    YAxis,
}

export default Axis;
