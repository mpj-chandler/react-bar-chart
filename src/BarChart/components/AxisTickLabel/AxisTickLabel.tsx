import React, { useRef, useState, useEffect, RefObject } from 'react';
import styles from './AxisTickLabel.scss';
import Axis from '../../enums/Axis';

export interface AxisTickLabelProps {
    label: string;
    xPos: number;
    yPos: number;
    axis: Axis;
}

const AxisTickLabel: React.FC<AxisTickLabelProps> = (props: AxisTickLabelProps) => {
    const ref = useRef();
    const [ xTransform, setXTransform ] = useState(0);
    const [ yTransform, setYTransform ] = useState(0);

    useEffect(() => {
        if (props.axis === Axis.XAxis) {
            setXTransform(-0.5 * (ref.current as unknown as SVGTextContentElement).getBBox().width);

            return;
        }
        setXTransform(-1.25 * (ref.current as unknown as SVGTextContentElement).getBBox().width);
        setYTransform(0.25 * (ref.current as unknown as SVGTextContentElement).getBBox().height);
    },
    [ref.current]);

    return (
        <text
                ref={ref as unknown as RefObject<SVGTextElement>}
                key={`tickText-${props.label}`}
                className={styles.AxisTickLabel}
                x={`${props.xPos}%`}
                y={`${props.yPos}%`}
                stroke={'black'}
                style={{ transform: `translate(${xTransform}px, ${yTransform}px)`}}
        >
            {props.label}
        </text>
    );
};

export default AxisTickLabel;
