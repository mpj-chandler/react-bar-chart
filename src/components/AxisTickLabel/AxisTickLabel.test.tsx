import React, { ReactElement } from 'react';
import TestRenderer, { TestRendererOptions } from 'react-test-renderer';
import AxisTickLabel from './AxisTickLabel';
import Axis from '../../enums/Axis';
import { ProgressPlugin } from 'webpack';


describe('Axis Text Label', () => {

    const label: string = '123456';

    const mockGetBBox = jest.fn().mockImplementation(() => {
        return {
            width: label.toString().length,
        };
    });

    const nodeMock = (element: ReactElement) => {
        if (element.type === 'text') {
            return {
                getBBox: () => {
                    return {
                        width: element.props.children[0].length,
                    };
                },
            };
        }

        return null;
    };


    describe('When mounted on an XAxis', () => {
        afterEach(() => {
            jest.resetAllMocks();
        });

        it('it renders consistently', () => {
            const component = TestRenderer.create(
                <AxisTickLabel
                    label={label}
                    xPos={0}
                    yPos={0}
                    axis={Axis.XAxis}
                />,
                { createNodeMock: nodeMock },
            );

            expect(component.toJSON()).toMatchSnapshot();
        });

        it('it adjusts to its text size', () => {

            let component = TestRenderer.create(
                <AxisTickLabel
                    label={label}
                    xPos={0}
                    yPos={0}
                    axis={Axis.XAxis}
                />,
                { createNodeMock: nodeMock },
            );

            const oldTransform = component.root.findByType('text').props.style.transform;

            component = TestRenderer.create(
                <AxisTickLabel
                    label={'123456789'}
                    xPos={0}
                    yPos={0}
                    axis={Axis.XAxis}
                />,
                { createNodeMock: nodeMock },
            );

            const newTransform = component.root.findByType('text').props.style.transform;

            expect(oldTransform).not.toStrictEqual(newTransform);
        });
    });

    describe('When mounted on a YAxis', () => {
        afterEach(() => {
            jest.resetAllMocks();
        });

        it('it renders consistently', () => {
            const component = TestRenderer.create(
                <AxisTickLabel
                    label={label}
                    xPos={0}
                    yPos={0}
                    axis={Axis.YAxis}
                />,
                { createNodeMock: nodeMock },
            );

            expect(component.toJSON()).toMatchSnapshot();
        });

        it('it adjusts to its text size', () => {

            let component = TestRenderer.create(
                <AxisTickLabel
                    label={label}
                    xPos={0}
                    yPos={0}
                    axis={Axis.YAxis}
                />,
                { createNodeMock: nodeMock },
            );

            const oldTransform = component.root.findByType('text').props.style.transform;

            component = TestRenderer.create(
                <AxisTickLabel
                    label={'123456789'}
                    xPos={0}
                    yPos={0}
                    axis={Axis.XAxis}
                />,
                { createNodeMock: nodeMock },
            );

            const newTransform = component.root.findByType('text').props.style.transform;

            expect(oldTransform).not.toStrictEqual(newTransform);
        });
    });
});
