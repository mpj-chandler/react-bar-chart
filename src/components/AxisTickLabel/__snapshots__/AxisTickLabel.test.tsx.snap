// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axis Text Label When mounted on a YAxis it renders consistently 1`] = `
<text
  className="AxisTickLabel"
  stroke="black"
  style={
    Object {
      "transform": "translate(-1.25px, NaNpx)",
    }
  }
  x="0%"
  y="0%"
>
  123456
</text>
`;

exports[`Axis Text Label When mounted on an XAxis it renders consistently 1`] = `
<text
  className="AxisTickLabel"
  stroke="black"
  style={
    Object {
      "transform": "translate(0px, 0px)",
    }
  }
  x="0%"
  y="0%"
>
  123456
</text>
`;
